lib/src/rules/avoid_runtimeType_toString.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Avoid calling toString() on runtimeType.';
+
+const _details = r'''
+
+Calling `toString` on a runtime type is a non-trivial operation that can
+negatively impact performance. It's better to avoid it.
+
+**BAD:**
+```
+class A {
+  String toString() => '$runtimeType()';
+}
+```
+
+**GOOD:**
+```
+class A {
+  String toString() => 'A()';
+}
+```
+
+This lint has some exceptions where performance is not a problem or where real
+type information is more important than performance:
+
+* in assertion
+* in throw expressions
+* in catch clauses
+* in mixin declaration
+* in abstract class
+
+''';
+
+class AvoidRuntimeTypeToString extends LintRule implements NodeLintRule {
+  AvoidRuntimeTypeToString()
+      : super(
+            name: 'avoid_runtimeType_toString',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addInterpolationExpression(this, visitor);
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (_canSkip(node)) {
+      return;
+    }
+    if (node.methodName.name == 'toString' &&
+        _isRuntimeTypeAccess(node.target)) {
+      rule.reportLint(node.target);
```

Comment:
I'd prefer having this diagnostic associated with `node.methodName`.

Reply:
Actually the issue is with the usage of `runtimeType` and I think it's better to associated this diagnotic with `runtimeType` like in the other case ( `'$runtimeType'`)

Reply:
> Actually the issue is with the usage of `runtimeType` ...

I don't believe that. :-) First, the description says:

> Calling `toString` on a runtime type ...

which clearly implies that the problem is with invoking `toString` on an instance of `Type`. Second, we're explicitly testing for `toString`. If the problem were with `runtimeType` then we wouldn't care how the result was being used.

Reply:
Done